src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -2970,15 +2968,20 @@ void ApplicationManagerImpl::UnregisterApplication(
 
     (hmi_capabilities_->get_hmi_language_handler())
         .OnUnregisterApplication(app_id);
-    AppV4DevicePredicate finder(handle);
-    ApplicationSharedPtr app = FindApp(applications(), finder);
-    if (!app) {
-      LOG4CXX_DEBUG(
-          logger_, "There is no more SDL4 apps with device handle: " << handle);
 
-      RemoveAppsWaitingForRegistration(handle);
-      RefreshCloudAppInformation();
-      SendUpdateAppList();
+    if (connection_handler().GetDeviceID(app_to_remove->mac_address(),
+                                         &handle)) {
```

Comment:
I'm a bit worried that there might be a case where GetDeviceID would fail for a cloud application, in which case the cloud device would not be properly recreated with `RefreshCloudAppInformation` maybe have an `else if (app_to_remove->is_cloud_app())` to cover this?